### PR TITLE
[SERF-1676] Add Optional Configurations And Fix Unwanted Error

### PR DESCRIPTION
**Pull request by laynax**
## Description

[SERF-1676] Instantiate configuration even if some app configuration files doesn't exist

* introduced configurable type for pre-defined apps
* renamed pkg/internal/configuration to pkg/configration/builder
* integrated all configs along with test files into pkg/configuration/apps
* making config files more similar and deleted settings for incompatibility

## Testing considerations

docker-compose run --rm sdk mage test:run

## Related links

* [SERF-1676](https://scribdjira.atlassian.net/browse/SERF-1676)

[commit messages]: https://chris.beams.io/posts/git-commit/

[SERF-1676]: https://scribdjira.atlassian.net/browse/SERF-1676?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Neurostep — COMMENTED** [review]
This is huge work, thanks for putting so much effort 🥇 But the changeset is huge and it is very hard to follow why we did particular changes.

I did the first pass and left some comments 👀 

I have a few major questions/comments that we have to address/discuss first:

1. Do we really need all of these changes? I see a massive work done here and I can see lots of good thinking put there. But in this PR, instead of fixing the issue we have, we are mixing a ton of refactoring work. That refactoring work is not very clearly described in the Pull Request. Can't we separate this PR into several: one would contain the same code that we currently have and solve the problem with `partial` configurations first; another would contain refactoring. wdyt?
2. Changes contain lots of breaking changes that would require service updates. Although, the whole idea of the task is to change how we interpret the configuration and `NewConfig` function, in this PR we do much more than that.
3. The actual code change has been made with one commit, it is really hard to follow it. Can't we do more granular commits that would touch on a particular change, especially refactoring?

Overall, I see a huge work done here and it is indeed appreciated 👍  But I have a feeling that we would benefit more if we scope changes around `NewConfig` first, and make it more flexible, This would fix the current problem we have. The refactoring part is, probably, better put into the separate PR and create a separate Jira for that where we would describe why we need this.

**Comment by laynax** [comment]
cc @Neurostep 
first of, thank you so much for putting time to read my PR and leave insight comments. I will explain my trail of problems and solutions here so we can figure out the parts we can discuss or fix.

tl;dr
It was not possible to do it in multiple commits, as it would either fail in compiling or testing.

---------------------------

I.**Problem**:
handle errors gracefully when a file is missing.
**I.Solution**:
config files must be specified to ignore having error, so `NewConfig()` must accept pre defined structures. it was accepted upon 
 variadic parameters would be a proper solution in a short video call. so we will have inevitable breaking change here. Solution is to create a type for pre-defined structures and pass it optionally to `NewConfig()`
**I.Design**:
a regular app `NewConfig()` looks like this:
```
func NewConfig() (*Config, error) {
	config := &Config{}
	viperBuilder := cbuilder.New("database")
...
```
we see 2 dependencies to this function, Config and builder so the type would be something like `type Configurable func(builder.Builder, *Configuration) error
`. This would allow us to change builders default behavior for every app config before passing to it.To make every app config to adapt this type, we face problem II

**II.Problem**:
- filename was included inside apps `NewConfig()`, now beside that it should be in configuration package where it belongs, it raises a problem, breaking tests and preventing it to have custom testfile directory, like [here](https://github.com/scribd/go-sdk/blob/main/pkg/database/config_test.go#L43)
- as long as we're supposed to extract builder out of app configs, best place for it is configuration package. now if we use viperBuilder, it wouldn't be forward compatible (we are forcing client to only use viperBuilder), can't apply different builders to different app configs and it's impossible to mock builder within `configuration` package.
- not a related to this ticket but still a miner problem, viperBuilder itself cannot be mocked for testing purposes.

**II.Solution**:
making viperBuilder an implementation of a `Builder` type as an interface{} could address all problems above.
**II.Design**:
the design is plain simple and is a copy paste of already written viperBuilder functionality (thanks to the author). To be able to write tests now, we encounter "III.challenge"

**III.Challenge***:
to test apps `NewConfig()` we need a mock builder to be created first and then pass it to `FetchConfig`. the challenge comes in 2 parts. first is that we should call this new "NewMockBuilder" for each test in each package which results in having same code in every package. Other miner challenge we have is that configuration package imports 7 internal packages which increases coupling.
**III.Solution**:
if we gather all config files to a separate package under configuration, both problems would be solved, now we have a `apps_test.go` that prepares a mock builder and is used in other app configs. this way, we have a more structured configuration files using same mocked builder to reduce entropy. also, now our configuration package is only importing 2 internal library, builder and apps. also by using same mocked builder, now we can have all test files under same directory and have better readability.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -140,4 +140,4 @@ jobs:
         name: Run tests
         with:
           image: ${{ github.repository }}:${{ env.GITHUB_SHA_SHORT }}
-          run: mage test:run
+          run: APP_ROOT=/sdk mage -v test:run
```
[the remark this answers is not in the document]
Comment (laynax) [reply .github/workflows/ci.yml:143]: I had errors in ci/test which I didn't have it in local test cases and only way to see the error from pipeline was to verbose it (conversation can be continued in other PR I have open regarding this).
about `APP_ROOT` we can have the discussion in other comments.

File: README.md [hunk 2]
```diff
@@ -83,63 +83,51 @@ variables on the fly.
...
-  `config/database.yml` configuration file.
-* `Redis`, containing the application Redis configuration, expects a
-  `config/redis.yml` configuration file. (TBD)
-* `PubSub`, containing the application pubsub configuration, expects a
-  `config/pubsub.yml` configuration file.
-* <insert new configuration here>
+```go
+    // import sdkconfig "github.com/scribd/go-sdk/pkg/configuration"
+    // equivalent of sdkconfig.NewConfigWithPath("config", ...)
+	config, err := sdkconfig.NewConfig(
```
[the remark this answers is not in the document]
Comment (laynax) [reply README.md:94]: I have explained why breaking changes was inevitable and we can discuss it if you would like to. but, if you think having breaking change is unnecessary, and might be problematic, I propose having another PR with a minor version update.

File: README.md [hunk 3]
```diff
@@ -562,13 +516,8 @@ var (
...
 func main() {
-	if loggerConfig, err := sdklogger.NewConfig(); err != nil {
-		log.Fatalf("Failed to initialize SDK logger configuration: %s", err.Error())
-	}
-
-	if trackingConfig, err := sdktracking.NewConfig(); err != nil {
-		log.Fatalf("Failed to initialize SDK tracking configuration: %s", err.Error())
-	}
+	loggerConfig := apps.Logger{}
+	trackingConfig := apps.Tracking{}
```
[the remark this answers is not in the document]
Comment (laynax) [reply README.md:520]: well I didn't mean to pass empty struct, just wanted to show the type it accepts. any change suggestion?

File: README.md [hunk 4]
```diff
@@ -1267,8 +1215,8 @@ tags to attach to every metric emitted by the client.
 
 This client is configured by default with:
 
-- `service:$APP_NAME`
-- `env:$ENVIRONMENT`
+- `service: $APP_NAME`
+- `env: $APP_ENV`
```
[the remark this answers is not in the document]
Comment (laynax) [reply README.md:1219]: I changed `ENVIRONMENT` to `APP_ENV` as we never had `ENVIRONMENT` in code, am I missing something?

File: README.md [hunk 5]
```diff
@@ -1299,11 +1247,7 @@ func main() {
 	applicationEnv := "development"
 	applicationName := "go-sdk-example"
 
-	metricsConfig := &metrics.Config{
-		Environment: applicationEnv,
-		App:         applicationName,
-	}
-	client, err := metrics.NewBuilder(metricsConfig).Build()
+	client, err := metrics.NewBuilder(applicationEnv, applicationName).Build()
```
[the remark this answers is not in the document]
Comment (laynax) [reply README.md:1250]: some apps including this didn't follow the same pattern as others, also [this](https://github.com/scribd/go-sdk/blob/main/pkg/metrics/config.go) didn't include test file. now by doing this, we save up a config.go and config_test.go from go-sdk, 4 lines in [go-chassis](https://github.com/scribd/go-chassis/blob/main/cmd/server/main.go#L69) and more simplicity in metrics system.
it is breaking change, yes, but I didn't mind it as long as we are already having a major update on configuration package.

File: docker-compose.yml [hunk 6]
```diff
@@ -6,6 +6,9 @@ services:
       context: .
       target: development
     image: microservices/sdk/go-sdk
+    environment:
+      APP_ENV: test
+      APP_ROOT: /sdk
```
[the remark this answers is not in the document]
Comment (laynax) [reply docker-compose.yml:11]: about env we will discuss, but I'm not sure what we broke.

File: pkg/configuration/apps/apps_test.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package apps
```
[the remark this answers is not in the document]
Comment (laynax) [reply pkg/configuration/apps/apps_test.go:1]: explained in long comment. any suggestion for better naming? this is the best I came up with :(

File: pkg/configuration/apps/tracking.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	if err != nil {
+		return err
+	}
+
+	if err = config.Unmarshal(t); err != nil {
+		return fmt.Errorf("Unable to decode into struct: %w", err)
+	}
+
+	t.Environment = config.String("ENV")
+	// TODO os must be fetched from config not os.
```
[the remark this answers is not in the document]
Comment (laynax) [reply pkg/configuration/apps/tracking.go:31]: so the idea behind it was to have only one package that uses `os.Getenv` (preferably configuration), this way we avoid loading them in every app config and have no validation over it.

File: pkg/configuration/builder/viper.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+// viperBuilder is a Builder to streamline Viper configuration and building.
+// implements builder.Builder.
+type viperBuilder struct {
+	vConf *viper.Viper
+	name  string
+}
+
+// NewViper initializes and returns a new ViperBuilder.
+// configDir is the reletive path for configuration directory.
+func NewViper(configDir, appName, appEnv, appRoot string) (Builder, error) {
```
[the remark this answers is not in the document]
Comment (laynax) [reply pkg/configuration/builder/viper.go:24]: I doubt that relates to our case. quote from mentioned doc:
> The implementing package should return concrete (usually pointer or struct) types
in our case, interface and it's implementation live under same package. AOF it should return interface so that if we change interface itself, the implementation requires change and raise compile error on correct line.
other reason, if we return a viper builder to `NewConfig`, within it we are able to make changes using viper which is not forward compatible (let's say we want to move on from viper, then that requires change on 2 packages).
we can have a chat if I couldn't deliver my thoughts here.

File: pkg/configuration/configurable.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
+package configuration
```
[the remark this answers is not in the document]
Comment (laynax) [reply pkg/configuration/configurable.go:1]: I'm not sure I understand it, mind elaborate more?

File: pkg/configuration/configuration.go [hunk 11]
```diff
@@ -1,72 +1,93 @@
...
-	appConfig, err := app.NewDefaultConfig()
-	if err != nil {
-		return config, err
-	}
+// NewConfigWithPath returns a new Config instance within a directory.
+// configDir is relative path of config directory.
+func NewConfigWithPath(confDir string, configurables ...Configurable) (*Configuration, error) {
+	appName := getAppNameMust()
+	appRoot := getAppRootMust()
+	appEnv := getEnvMust()
```
[the remark this answers is not in the document]
Comment (laynax) [reply pkg/configuration/configuration.go:45]: app name and env was used by many app configs using `os.Getenv`, this way we can have them all set together and also can put some validation on top of it. now why did I fatal instead of returning error, is because I believe this error is different than others in 2 sense. first is that this is the first line of initializing configuration which is the first line of your main func, one of very first needs of program so make sense to fatal without need to see the log or trace (considering the is no logger too). the other is that this error is not coding related, so must be plain text that anyone from monitoring/devops team could get to it fast and easy.
what do you think?

File: pkg/instrumentation/database_test.go [hunk 12]
```diff
@@ -74,7 +74,7 @@ func TestInstrumentDatabase(t *testing.T) {
 	mt := mocktracer.Start()
 	defer mt.Stop()
 
-	dbFile := "/tmp/test_db"
+	dbFile := "/var/tmp/test_db"
```
[the remark this answers is not in the document]
Comment (laynax) [reply pkg/instrumentation/database_test.go:77]: https://github.com/scribd/go-sdk/runs/7235016120?check_suite_focus=true
I already have a fix ready for here in another PR. we should use `t.TempDir()`.
avoided doing that to make this PR simpler (which I failed XD)
